lib/common/operations.c

```diff
@@ -65,164 +62,119 @@ convert_interval(const char *s, guint *interval_ms)
+        // Must have at least X_MATCH before position
+        if (position > (match_len + 1)) {
+            const size_t possible = position - match_len - 1;
 
-            if (op_type) {
-                free(*op_type);
-                *op_type = NULL;
+            if ((key[possible] == '_')
+                && (strncmp(key + possible + 1, matches[i], match_len) == 0)) {
+                return possible;
             }
-
-            return FALSE;
         }
     }
-
-    return TRUE;
+    return 0;
 }
 
 gboolean
 parse_op_key(const char *key, char **rsc_id, char **op_type, guint *interval_ms)
 {
-    char *underbar1 = NULL;
-    char *underbar2 = NULL;
-    char *underbar3 = NULL;
+    guint local_interval_ms = 0;
+    const size_t key_len = (key == NULL)? 0 : strlen(key);
+
+    // Operation keys must be formatted as RSC_ACTION_INTERVAL
+    size_t action_underbar = 0;   // Index in key of underbar before ACTION
+    size_t interval_underbar = 0; // Index in key of underbar before INTERVAL
+    size_t possible = 0;
+
+    /* Underbar was a poor choice of separator since both RSC and ACTION can
+     * contain underbars. Here, list action names and name prefixes that can.
+     */
+    const char *actions_with_underbars[] = {
+        "migrate_from",
+        "migrate_to",
+        NULL
+    };
+    const char *action_prefixes_with_underbars[] = {
+        "pre_notify",
+        "post_notify",
+        NULL,
+    };
 
     // Initialize output variables in case of early return
     if (rsc_id) {
         *rsc_id = NULL;
     }
-
     if (op_type) {
         *op_type = NULL;
     }
-
     if (interval_ms) {
         *interval_ms = 0;
     }
 
-    CRM_CHECK(key && *key, return FALSE);
-
-    underbar1 = strchr(key, '_');
-    if (!underbar1) {
+    // RSC_ACTION_INTERVAL implies a minimum of 5 characters
+    if (key_len < 5) {
         return FALSE;
     }
 
-    underbar2 = strchr(underbar1+1, '_');
-    if (!underbar2) {
+    // Find, parse, and validate interval
+    interval_underbar = key_len - 2;
+    while ((interval_underbar > 2) && (key[interval_underbar] != '_')) {
+        --interval_underbar;
+    }
+    if ((interval_underbar == 2)
+        || !convert_interval(key + interval_underbar + 1, &local_interval_ms)) {
         return FALSE;
     }
 
-    underbar3 = strchr(underbar2+1, '_');
+    // Find the base (OCF) action name, disregarding prefixes
+    action_underbar = match_before(key, interval_underbar,
+                                   actions_with_underbars);
+    if (action_underbar == 0) {
+        action_underbar = interval_underbar - 2;
+        while ((action_underbar > 0) && (key[action_underbar] != '_')) {
+            --action_underbar;
+        }
+        if (action_underbar == 0) {
+            return FALSE;
+        }
+    }
+    possible = match_before(key, action_underbar,
+                            action_prefixes_with_underbars);
+    if (possible != 0) {
+        action_underbar = possible;
+    }
```

Comment:
Yeah, I think RSC@ACTION_INTERVAL would work (since @ is not allowed in XML IDs, and RSC can already contain a colon). Switching over should be doable, we just need to base the format on the DC version when recording history (and allowing either format when reading history). Probably worth waiting until a new major/minor series release though, and it might even be possible to use an XSL transform to upgrade old history.